### PR TITLE
[App Config] Add support for exporting App Configuration references to App Service

### PR DESCRIPTION
**Pull request by albertofori**
**Related command**
`az appconfig kv import`
`az appconfig kv export`

**Description**
1. Adds support for exporting app configuration settings as references to app service. Can be done by adding the flag `--export-as-reference` or `-r` when exporting to an app service destination.
2. App Configuration references are skipped when importing configuration settings from app service to App Configuration.

**Testing Guide**

**History Notes**

[App Config] `az appconfig kv export`: Support exporting app configuration settings as references to App Service
[App Config] `az appconfig kv import`: Ensure app configuration references are not imported from App Service

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
App Config

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 1]
```diff
@@ -281,6 +281,19 @@ def __write_kv_and_features_to_file(file_path, key_values=None, features=None, f
         raise CLIError("Failed to export key-values to file. " + str(exception))
 
 
+# Exported in the format @Microsoft.AppConfiguration(Endpoint=<storeEndpoint>; Key=<kvKey>; Value=<kvValue>; Label=<kvLabel>).
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:284]: We shouldn't export any 'Value' in the reference

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 2]
```diff
@@ -518,10 +531,12 @@ def __serialize_kv_list_to_comparable_json_object(keyvalues, level):
...
         for kv in keyvalues:
             kv_json = {'value': kv.value}
+            # Explicitly assign slot settings for comparison.
+            slot_setting = 'false'
             if kv.tags:
-                for tag_k, tag_v in kv.tags.items():
-                    if tag_k == 'AppService:SlotSetting':
-                        kv_json[tag_k] = tag_v
+                slot_setting = kv.tags.get('AppService:SlotSetting', 'false')
+            kv_json['AppService:SlotSetting'] = slot_setting
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:538]: Should we move this tag to the constants file too?

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 1 above)
Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:284]: I see you're not really adding value in the reference. Can you update the comment?

File: src/azure-cli/azure/cli/command_modules/appconfig/_params.py [hunk 3]
```diff
@@ -189,6 +189,7 @@ def load_arguments(self, _):
 
     with self.argument_context('appconfig kv export', arg_group='AppService') as c:
         c.argument('appservice_account', validator=validate_appservice_name_or_id, help='ARM ID for AppService OR the name of the AppService, assuming it is in the same subscription and resource group as the App Configuration. Required for AppService arguments')
+        c.argument('export_as_reference', options_list=['--export-as-reference', '-r'], help="Export key-values as App Configuration references (Preview). For more information, see https://docs.microsoft.com/en-us/azure/app-service/app-service-configuration-references ")
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_params.py:192]: We can add preview tag to this argument (no need to add 'Preview' in the help text).

```
c.argument('export_as_reference', is_preview=True, .....)
```

Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_params.py:192]: Another thing we can add is the three state flag:
` c.argument('export_as_reference', is_preview=True, arg_type=get_three_state_flag(), .......)`

This enables users to provide true/false/empty values to this argument. You can look at `--enable-public-network` argument for sample usage.

File: src/azure-cli/azure/cli/command_modules/appconfig/_utils.py [hunk 4]
```diff
@@ -132,11 +132,16 @@ def is_valid_connection_string(connection_string):
...
+    return None
+
+
+def get_store_endpoint_from_connection_string(connection_string):
     if is_valid_connection_string(connection_string):
         segments = dict(seg.split("=", 1) for seg in connection_string.split(";"))
-        endpoint = segments.get("Endpoint")
-        if endpoint:
-            return endpoint.split("//")[1].split('.')[0]
+        return segments.get("Endpoint")
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_utils.py:144]: Can we return `segments.get("Endpoint", None)`?

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 5]
```diff
@@ -281,6 +281,19 @@ def __write_kv_and_features_to_file(file_path, key_values=None, features=None, f
         raise CLIError("Failed to export key-values to file. " + str(exception))
 
 
+# Exported in the format @Microsoft.AppConfiguration(Endpoint=<storeEndpoint>; Key=<kvKey>; Value=<kvValue>; Label=<kvLabel>).
+# Label is optional
+
+def __map_to_appservice_config_reference(key_value, endpoint):
+    if endpoint is None:
+        raise CLIError("No endpoint provided.")
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:289]: Instead of raising error, we can log a warning and continue exporting without converting it to reference. This decision can be made even before __map_to_appservice_config_reference is invoked. You are already resolving store endpoint, if you just add a None check over there, we can log warning and export as regular key-values. Thoughts?

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 6]
```diff
@@ -281,6 +281,19 @@ def __write_kv_and_features_to_file(file_path, key_values=None, features=None, f
...
+
+    label = key_value.label
+    key_value.value = AppServiceConstants.APPSVC_CONFIG_REFERENCE_PREFIX + 'Endpoint={0}; Key={1}'.format(
+        endpoint, key_value.key) + ('; Label={0}'.format(label) if label is not None else '') + ')'
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:293]: `key_value.key` has already been trimmed of the prefix that user might have specified with `--prefix` argument (we do this prefix trimming in `__read_kv_from_config_store`). 
While the AppSetting name is not impacted if the prefix is trimmed, this reference will not work since the trimmed key is not present in AppConfig.

I think this should be the behavior when --prefix and -r both are provided:

Lets say the prefix is "Beta_".

AppConfig key-value:
Key: Beta_Feature
Value: Test

After exporting to AppService:
Key: Feature
Value: @Microsoft.AppConfiguration(Endpoint=mystore.azconfig.io; Key=Beta_Feature)

Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:293]: The simplest solution I can think of is pass the prefix to this method and append it to the value string appropriately.

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 7]
```diff
@@ -416,19 +429,19 @@ def __read_kv_from_app_service(cmd, appservice_account, prefix_to_add="", conten
             cmd, resource_group_name=appservice_account["resource_group"], name=appservice_account["name"], slot=slot)
         for item in settings:
             key = prefix_to_add + item['name']
-            if validate_import_key(key):
+            value = item['value']
+            if validate_import_key(key) and not value.strip().lower().startswith(AppServiceConstants.APPSVC_CONFIG_REFERENCE_PREFIX.lower()):  # Exclude app configuration references.
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:433]: Can we add a warning log for every AppConfig reference that we are skipping during import? Users may not be expecting it so its better to be more explicit.

File: src/azure-cli/azure/cli/command_modules/appconfig/_params.py (hunk 3 above)
Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_params.py:192]: Can you also add a `validator` for this argument?
I think we should throw an error if this argument is used with any destination other than AppService. You can look at this validator for a sample: https://github.com/Azure/azure-cli/blob/f7479671e841d196d3312523c302bab853eb25c2/src/azure-cli/azure/cli/command_modules/appconfig/_validators.py#L226

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 8]
```diff
@@ -281,6 +281,19 @@ def __write_kv_and_features_to_file(file_path, key_values=None, features=None, f
...
 
 
+# Exported in the format @Microsoft.AppConfiguration(Endpoint=<storeEndpoint>; Key=<kvKey>; Value=<kvValue>; Label=<kvLabel>).
+# Label is optional
+
+def __map_to_appservice_config_reference(key_value, endpoint):
+    if endpoint is None:
+        raise CLIError("No endpoint provided.")
+
+    label = key_value.label
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:291]: What is the exact value of `key_value.label` when a key-value has no label? Is it `None`?

File: src/azure-cli/azure/cli/command_modules/appconfig/_utils.py (hunk 4 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_utils.py:144]: `.get` returns `None` by default. Should we still explicitly state it?

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 5 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:289]: You make a good point. Come to think of it, the None check is redundant since we would have already fetched `src_kvs`. That would mean the endpoint would have been resolved successfully else an exception would have been thrown when trying to create `azconfig_client`.
Could I be missing something here? 
If not, would you suggest we consider this case?

File: src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py [hunk 9]
```diff
@@ -249,6 +251,12 @@ def export_config(cmd,
     if skip_keyvault:
         src_kvs = [keyvalue for keyvalue in src_kvs if keyvalue.content_type != KeyVaultConstants.KEYVAULT_CONTENT_TYPE]
 
+    if destination == 'appservice' and export_as_reference:
+        if endpoint is None:
+            endpoint = get_store_endpoint_from_connection_string(connection_string) or resolve_store_metadata(cmd, name)[1]
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py:256]: This is where I was thinking we can validate if an endpoint has been resolved or not:
```
endpoint = get_store_endpoint_from_connection_string(connection_string) or resolve_store_metadata(cmd, name)[1]
if endpoint:
     src_kvs = [__map_to_appservice_config_reference(kv, endpoint) for kv in src_kvs]
else:
    logger.warning("Cannot export as AppConfig references because store endpoint could not be resolved. Key-values will be exported as-is.")
```

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 5 above)
Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:289]: See my other comment for this.

Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:289]: See this comment: https://github.com/Azure/azure-cli/pull/23795#discussion_r977926226

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 1 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py]: Thanks for pointing this out. pushed an update to address this.

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 6 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:290]: Since trimming will check whether the prefix exists before removing it, I move the trimming of the prefix to the `__map_to_appservice_config_reference` when the `export_as_reference` flag is indicated.

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 7 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py]: Update has been pushed for this. Kindly let me know if the wording works.

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py (hunk 8 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:288]: Yes, it is.

File: src/azure-cli/azure/cli/command_modules/appconfig/_params.py (hunk 3 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_params.py]: Update pushed for this.

File: src/azure-cli/azure/cli/command_modules/appconfig/_constants.py [hunk 10]
```diff
@@ -16,7 +16,12 @@ class FeatureFlagConstants:
 
 class KeyVaultConstants:
     KEYVAULT_CONTENT_TYPE = "application/vnd.microsoft.appconfig.keyvaultref+json;charset=utf-8"
+
+
+class AppServiceConstants:
+    APPSVC_CONFIG_REFERENCE_PREFIX = "@Microsoft.AppConfiguration("
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_constants.py:22]: I just noticed that APPSVC_KEYVAULT_PREFIX doesn't end with `(`. Can we have AppConfig prefix without the ending bracket for consistency?

File: src/azure-cli/azure/cli/command_modules/appconfig/tests/latest/test_appconfig_commands.py [hunk 11]
```diff
@@ -937,6 +937,73 @@ def test_appconfig_to_appservice_import_export(self, resource_group, location):
         slot = self.create_random_name(prefix='Slot', length=24)
         self.cmd('webapp deployment slot create -g {} -n {} -s {}'.format(resource_group, webapp_name, slot))
 
+        # App configuration reference tests
+        # Create new key-values
+        entry_key = 'TestKey'
+        entry_value = 'TestValue'
+        entry_label = 'AppServiceReferenceExport'
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/tests/latest/test_appconfig_commands.py:944]: Do you mind adding a test for null label too? Just to make sure that the reference string is formed as expected.

File: src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py [hunk 12]
```diff
@@ -243,12 +245,19 @@ def export_config(cmd,
...
                                           key=key,
                                           label=label if label else SearchFilterOptions.EMPTY_LABEL,
-                                          prefix_to_remove=prefix,
+                                          prefix_to_remove=prefix if not export_as_reference else "",
                                           cli_ctx=cmd.cli_ctx if resolve_keyvault else None)
 
     if skip_keyvault:
         src_kvs = [keyvalue for keyvalue in src_kvs if keyvalue.content_type != KeyVaultConstants.KEYVAULT_CONTENT_TYPE]
 
+    if destination == 'appservice' and export_as_reference:
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py:254]: We can move this block after we have invoked `__discard_features_from_retrieved_kv`. That's because we dont want to call `__map_to_appservice_config_reference` for any feature flags.

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 13]
```diff
@@ -281,6 +281,19 @@ def __write_kv_and_features_to_file(file_path, key_values=None, features=None, f
...
+# Label is optional
+
+def __map_to_appservice_config_reference(key_value, endpoint, prefix):
+    label = key_value.label
+    key_value.value = AppServiceConstants.APPSVC_CONFIG_REFERENCE_PREFIX + 'Endpoint={0}; Key={1}'.format(
+        endpoint, key_value.key) + ('; Label={0}'.format(label) if label is not None else '') + ')'
+
+    if key_value.key.startswith(prefix):
+        key_value.key = key_value.key[len(prefix):]
+    return key_value
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:294]: In `__write_kv_to_app_service`, we check if the content type is that of KeyVault reference and try to convert the value to AppService format KeyVault reference. But if we have already converted a KeyVault reference to an AppConfig reference here, the check in  `__write_kv_to_app_service` is unnecessarily going to raise an exception while deserializing the value. Although the exception is not thrown and the value is treated as regular string value, we should avoid doing things that will definitely raise exceptions, because catching an exception is expensive.

Can you check if there are any side effects if we clear the content type for all key-values here? I dont think we need the content type while exporting to AppService after this point.

Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:294]: I confirmed that the content type is not used when exporting to app service. So this can be changed without any side effects. Technically, the key-value is no more a key-vault reference.  

Another option would be to supply the `export-as-reference` flag to the `__write_kv_to_app_service` method and skip the key-vault reference validation when it is true.

Comment (avanigupta) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:294]: I'd prefer changing the content type, because as you said - its technically not a KeyVault reference anymore.

File: src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py [hunk 14]
```diff
@@ -251,21 +251,21 @@ def export_config(cmd,
...
-        return
+            __export_kvset_to_file(file_path=path, keyvalues=src_kvs, yes=yes)
+            return
+
+    if destination == 'appservice' and export_as_reference:
+            if endpoint is None:
+                # Endpoint will not be None as it is already resolved in creating azconfig_client
+                endpoint = get_store_endpoint_from_connection_string(connection_string) or resolve_store_metadata(cmd, name)[1]
+
+            src_kvs = [__map_to_appservice_config_reference(kv, endpoint, prefix) for kv in src_kvs]
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py:268]: Looks like an extra tab space got added to these lines

**Review by avanigupta — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
@albertofori Could you please resolve those CI issues?

**Comment by zhoxing-ms** [comment]
> [App Config] Change: az appconfig kv export: Support exporting app configuration settings as references to App Service
[App Config] Change: az appconfig kv import: Ensure app configuration references are not imported from App Service

For ordinary changes that are not breaking changes, we do not need to add the `Change` tag to the PR description/title.

File: src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py (hunk 14 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/keyvalue.py]: My bad! Just fixed it.

**Comment by albertofori** [comment]
> @albertofori Could you please resolve those CI issues?

@zhoxing-ms Please I just pushed a fix to address the issues, thanks!

**Review by zhoxing-ms — APPROVED** [review]
